docs/uri-parser-rewriter-proposal.md

```diff
@@ -0,0 +1,272 @@
+
+The public `QueryNode` struct is smaller than the internal `ExpressionNode` and would be cheaper to move around.
+
+`ExpressionNode` can represent terminal nodes like primitives and identifiers, binary operators with a left and right child
+and collections with an arbitrary number of children. It can also represent nodes with nested children.
+
+- The `Count` property represents the number of top-level children a node has. The `LastChild` is the index of the last child in the array.
+- We use `0` for the Count to represent terminal nodes
+- The first child will always be the at parent's index + 1.
+
+For nodes with non-nested children, we can easily find out in O(1):
+
+- whether it has children, via the `Count` property
+- the index of the child (`index + 1`)
+- the index of any child (since we know where each child item will be)
+
+However, if children are nested (e.g. an `or` node that has a left `and` node), it's harder to tell children are located. In this case,
+we have to iterate through each child and skip over nested children to enumerate the top-level children. We can set the most significant bit to 1
+to tell whether a node has "complex" or "nested" childs. If this bit is 0, we use the O(1) approach to find children by index, if the bit is 1 (i.e. `Count < 0`)
+we use the iterative approach.
+
+Let's use the following query to demonstrate how the array will be laid out:
+
+```text
+category in ('electronics', substring(field, 1), "technolgy")
+```
+
+Here's a tree representation of the query:
+
+![Tree representation of the query](uri-parser-nested-in-query-tree.png)
+
+**Note** This is not how the current `UriQueryExpressionParser` represents this tree. Notably, it does not parse the the collection operand of and the `in` operator.
+
+Here's how this would look like in the node array:
+
+![Array layout of the queyr](./uri-parser-nested-in-query-array.png)
+
+The downside of this approach is that there could be multiple nesting levels between two siblings of the same parent. That would require multiple hops of to skip over from one child to its next sibling. The number of hops is proportional to the level of nesting. If we have highly nested collections, this could noticeably slow down the iteration or indexing beyond O(n). If we anticipate highly nested collections, we could reduce the number of hops to one per sibling by storing more metadata about each child and level (TODO: demonstrate this).
+
+How does the end user process this tree? Since we're using structs, we can't provide a traditional visitor that's based on inheritance. The basic approach is for the user to simply traverse the tree using the APIs on the `QueryNode` struct, like they would any other DOM-like structure. But this may be tedious for some cases. We can propose a handler that exposes a different method based on the type of node for convenience, e.g.:
+
+```csharp
+interface IQueryNodeHandler<T>
+{   
+    public T HandleIntNode(QueryNode intLiteralNode);

```

Comment:
Is there a huge perf overhead for these methods to take in more specific types? Like, it would be great if we could make this `IntQueryNode` instead

Reply:
Also, we have to be very careful about this because we will need to constantly add support for `async`, `unsafe`, methods that require parameterized context, `allows ref struct`, etc.

Reply:
Based on your suggestions, I will investigate the cost of adding more specific types on top `QueryNode` for the different methods, e.g. `Handle(IntQueryNode)`

Where `IntQueryNode` could be defined as 

```c#
struct IntQueryNode
{
     private QueryNode _node;
     public int GetIntValue() => _node.GetIntValue();
}
```

Maybe it could be optimized to a ref struct itself or it could reference the expression parser directly if such optimizations prove meaningful and practical.